File: day11/main.py
from copy import deepcopy
from typing import Callable, List


def parse_input(filepath: str, split: str = "\n") -> list:
    with open(filepath, "r") as file:
        content = file.read()
    return content.split(split)


coords = ((1, -1), (1, 0), (1, 1), (0, -1), (0, 1), (-1, -1), (-1, 0), (-1, 1))


def countNeighbors(seats: list, y: int, x: int) -> int:
    occupied = 0
    for dy, dx in coords:
        if y + dy < 0:
            continue
        elif y + dy >= len(seats):
            continue
        if x + dx < 0:
            continue
        elif x + dx >= len(seats[0]):
            continue
        neighbor_seat = seats[y + dy][x + dx]
        if neighbor_seat == "#":
            occupied += 1
    return occupied


def countNeighborsSight(seats: list, y: int, x: int) -> int:
    occupied = 0
    ny = len(seats)
    nx = len(seats[0])
    for dy, dx in coords:
        i = 1
        while 0 <= y + dy * i < ny and 0 <= x + dx * i < nx:
            seat = seats[y + dy * i][x + dx * i]
            if seat == "#":
                occupied += 1
                break
            elif seat == "L":
                break
            i += 1
    return occupied


def solve(seats: List[List[str]], count_func: Callable, n_occupied: int) -> int:
    while True:
        newSeats = deepcopy(seats)
        # print(f"Iteration {i}")
        changed = 0
        for y, row in enumerate(seats):
            # print("".join(row), end="\t")
            for x, seat in enumerate(row):
                n = count_func(seats, y, x)
                # print(n, end="")
                if seat == "L" and n == 0:
                    # -> occupied
                    newSeats[y][x] = "#"
                    changed += 1
                    continue
                if seat == "#" and n >= n_occupied:
                    # -> empty
                    newSeats[y][x] = "L"
                    changed += 1
                    continue
            # print("")
        seats = deepcopy(newSeats)

        if changed == 0:
            break

    occupied = 0
    for row in seats:
        for seat in row:
            if seat == "#":
                occupied += 1

    return occupied


if __name__ == "__main__":
    lines = parse_input("./input")
    seats = []
    for line in lines:
        seats.append([seat for seat in line])

    sol1 = solve(seats, countNeighbors, 4)
    print(f"Exercise 1: {sol1} occupied seats.")
    sol2 = solve(seats, countNeighborsSight, 5)
    print(f"Exercise 2: {sol2} occupied seats.")

File: day11/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func readFileLines(filepath string, split string) []string {
	content, _ := ioutil.ReadFile(filepath)
	return strings.Split(string(content), split)
}

// const empty = 76
// const floor = 46
// const occupied = 35

func countNeighbors(seats [][]string, y int, x int) int {
	var occupiedNeighbors int
	ny := len(seats)
	nx := len(seats[0])

	var coords = [][]int{
		{0, -1}, {-1, 0}, {0, 1}, {1, 0},
		{1, -1}, {-1, -1}, {-1, 1}, {1, 1}}

	for _, coord := range coords {
		dy, dx := coord[0], coord[1]
		if y+dy < 0 {
			continue
		} else if y+dy >= ny {
			continue
		}
		if x+dx < 0 {
			continue
		} else if x+dx >= nx {
			continue
		}
		seat := seats[y+dy][x+dx]
		if seat == "#" {
			occupiedNeighbors++
		}
	}
	return occupiedNeighbors
}

func main() {
	lines := readFileLines("./test", "\n")
	var seats [][]string
	for y, row := range lines {
		seats = append(seats, []string{})
		for _, seat := range row {
			seats[y] = append(seats[y], string(seat))
		}
	}
	ny := len(seats)
	nx := len(seats[0])

	var nChanges = 1
	var i int

	for nChanges > 0 {
		var newSeats = make([][]string, nx, ny)
		copy(newSeats, seats)
		// fmt.Printf("\nIteration %d\n", i)
		i++
		nChanges = 0
		for y := 0; y < ny; y++ {
			// fmt.Printf("%v\t", seats[y])
			for x := 0; x < nx; x++ {

				nNeighbors := countNeighbors(seats, y, x)
				// fmt.Printf("%d", nNeighbors)
				seat := seats[y][x]
				if seat == "." {
					continue
				}
				if seat == "L" {
					if nNeighbors == 0 {
						// -> occupied
						newSeats[y][x] = "#"
						nChanges++
						continue
					}
				} else if seat == "#" {
					if nNeighbors >= 4 {
						// -> empty
						newSeats[y][x] = "L"
						nChanges++
						continue
					}
				}

			}
			// fmt.Printf("\n")
		}
		copy(seats, newSeats)

		// fmt.Println(nChanges)
	}
	var occupiedSeats int
	for y := 0; y < ny; y++ {
		for x := 0; x < nx; x++ {
			if seats[y][x] == "#" {
				occupiedSeats++
			}
		}
	}
	fmt.Println(occupiedSeats)

}
